File: Terrazas-Chavez-Jose-Alexis/Practicas/Python/practica_5.py
'''
Programa que crea un proceso y seguidamente crealiza una sincronización de PID con el proceso
subsecuente, gracias al comando "execl" a través de la libreria "os" , esto quiere decir que el
proceso del padre lo obtendra la efecución del comando ps -f (se coloa f y no fea para visualizar
rapidamente el PID) por lo que en ese momento se detendrá el código de ejecución y no imprime el 
print despues de la sincronización ya que se cambio el proceso.
'''
import os
import time


def main():

    print(f"Proceso padre con PID {os.getpid()}")

    # Realizar el fork

    #comando = "/bin/ps"  # Ruta absoluta al binario
    #argumentos = ["ps", "-fea"]

    # Ejecutar el comando con execl. ¡Reemplaza el proceso actual!
    os.execl("/bin/ps", "ps","-f")
    print("nunca me veran")
if __name__ == "__main__":
    main()

File: Terrazas-Chavez-Jose-Alexis/Practicas/C++/practica_5.cpp
/*
Programa que crea tres porcesos y realiza un cambio de contexto cada que uno termina,
el cambio de contexto lo determina el quantum.
Si se quiere detener proceso se tiene que colocar el comando ps -fea en otra terminar 
y una vez ubicado el PID matarlo con el el comando $kill <PID>
*/
#include <iostream>
#include <unistd.h>  // Para fork(), getpid(), sleep()
#include <cstdlib>   // Para exit()
using namespace std;

void proceso(int pid) {
    int i = 1;
    while (true) {
        for (int t = 0; t <= pid; ++t) {
            cout << "\t";
        }
        cout << "<" << getpid() << "," << i << ">" << endl;
        i++;
        // sleep(1);  // Simular el quantum con una pausa
    }
}

int main() {
    const int N = 3;  // Número de procesos

    for (int i = 1; i <= N; ++i) {
        pid_t pid = fork();  // Crear proceso hijo

        if (pid == 0) {
            proceso(i);  // Ejecutar función en proceso hijo
            exit(0);  // Terminar el proceso hijo
        }
    }

    // Mantener el proceso padre corriendo
    while (true) {
        sleep(1);  
    }

    return 0;
}
